src/blocked.c

```diff
@@ -325,7 +325,7 @@ void handleClientsBlockedOnKeys(void) {
      * (i.e. not from call(), module context, etc.) */
     serverAssert(server.also_propagate.numops == 0);
 
-    while(listLength(server.ready_keys) != 0) {
+    if (listLength(server.ready_keys) != 0) {
```

Comment:
the assertion in logreqres, It looks like it's caused by this line

Reply:
the test was introduced in #6798

Reply:
ok, i think i was wrong to change that.
it means that if a command being unblocked causes another command to get unblocked (like a BRPOPLPUSH would do), then the new unblocked command will wait for later to get processed rather than right away.
i.e. that change is wrong, and it's also not needed to fix the current problem (validated that using the original python scripts in the issue too).

what it also means is that:
1. we have no test that validates that (i.e. that bug i created went undetected).
2. the assertion in logreqres.c may still be wrong IMHO, bug we can leave it aside for now.

Reply:
yes, so i should revert the change, and then i will try to write a test, the same one like the module one, using BRPOPLPUSH

Reply:
i'm not sure it's a bug that you can test for... it'll not have any real implication since we do call handleClientsBlockedOnKeys in beforeSleep again, and redis will still behave correctly.

maybe the only way to detect it is to write a bunch of commands in a pipeline, and watch the order in which they're executed (e.g. propagated to replica or AOF).
with the original version, a BLPOP that's released by BRPOPLPUSH will be executed before the next command in the pipeline, and with the bug, it'll be executed after.

however, i'm still not sure how all of that is related to the logreqres assertion, and why it only happens with the module version.

Reply:
> however, i'm still not sure how all of that is related to the logreqres assertion, and why it only happens with the module version.

i maybe had some lucks, i can tell what i found:
if we change the while to if, we will trigger the handleClientsBlockedOnKeys in beforesleep

so in beforesleep, we call handleClientsBlockedOnKeys, then it call moduleUnblockClientOnKey, then it call moduleUnblockClient, it add the client to moduleUnblockedClients
back to beforesleep, we call handleClientsWithPendingWritesUsingThreads, it writes the data of buf to the client, so client->bufpos became 0
maybe in the next beforesleep, we call moduleHandleBlockedClients, then it call unblockClient, then it call reqresAppendResponse, trigger the assert. (because the bufpos is 0)

so i guess, maybe we need to move moduleHandleBlockedClients after handleClientsBlockedOnKeys
or add a new moduleHandleBlockedClients block after handleClientsBlockedOnKeys
or move handleClientsBlockedOnKeys to the front of moduleHandleBlockedClients block
i have verified that these methods will pass the test

Reply:
out of these, moving handleClientsBlockedOnKeys upwards (also above processUnblockedClients) seems the best to me.
but i don't think i know this area well enough, and i'd like @guybe7 to comment.
also:
1. i'm not certain if this is a real issue or just a reqres limitation.
2. other than remporarily messing up the while in handleClientsBlockedOnKeys, we don't have a way to reproduce this.
3. it's unrelated to this PR.

> i can try the pipeline way, do you want me to do that? (In other words, do we need such a test)
I'm still looking at the diffs in the module part, hopefully I can find something

i was about to say that i don't think we should bother, but considering the above nuances of ordering of things, i suppose more tests around blocking edge cases and how they are processed, can't hurt. the test seems not too complicated, so feel free to go ahead (can be part of this PR, and should include some comment about what it tests / why. i.e. correct order of command execution)

Reply:
@enjoy-binbin i discussed this with @guybe7 and he'll take over the part related to the module test, reqres assertion and beforeSleep task order.

do you wanna try writing the test we discussed above (correct command order execution, improved coverage) in this PR, or shall i merge it as is, and leave it for a future PR?

Reply:
i think we can merge this first, feels like the test is unrelated  to this PR, I can add it in the future

Reply:
btw I just saw your comment above.

> out of these, moving handleClientsBlockedOnKeys upwards (also above processUnblockedClients) seems the best to me.

that's what i thought too

I agree with those points (the 1, 2, 3)